pkg/ebpf/telemetry/errors_telemetry.go

```diff
@@ -43,213 +40,346 @@ var helperNames = map[int]string{
-		opts.MapEditors[probes.MapErrTelemetryMap] = b.mapErrMap.Map()
-	}
-	if b.helperErrMap != nil {
-		opts.MapEditors[probes.HelperErrTelemetryMap] = b.helperErrMap.Map()
+}
+
+// taken from: https://github.com/cilium/ebpf/blob/main/asm/opcode.go#L109
+// this returns the total number of bytes to encode the instruction
+func countRawBPFIns(ins *asm.Instruction) int {
+	if ins.OpCode.IsDWordLoad() {
+		return 2
 	}
+	return 1
 }
 
-// populateMapsWithKeys initializes the maps for holding telemetry info.
-// It must be called after the manager is initialized
-func (b *EBPFTelemetry) populateMapsWithKeys(m *manager.Manager) error {
-	b.mtx.Lock()
-	defer b.mtx.Unlock()
+// initializeProbeKeys assignes an integer key to each probe in the manager.
+// This key is the index in the array at which telemetry for this probe is kept
+// See pkg/ebpf/c/telemetry_types.h for the definition of the struct holding telemetry
+func initializeProbeKeys(programs map[string]*ebpf.ProgramSpec, bpfTelemetry *EBPFTelemetry) {
+	bpfTelemetry.mtx.Lock()
+	defer bpfTelemetry.mtx.Unlock()
 
-	// first manager to call will populate the maps
-	if b.mapErrMap == nil {
-		b.mapErrMap, _ = maps.GetMap[uint64, MapErrTelemetry](m, probes.MapErrTelemetryMap)
-	}
-	if b.helperErrMap == nil {
-		b.helperErrMap, _ = maps.GetMap[uint64, HelperErrTelemetry](m, probes.HelperErrTelemetryMap)
+	for fn := range programs {
+		bpfTelemetry.probeKeys[fn] = bpfTelemetry.probeIndex
+		bpfTelemetry.probeIndex++
 	}
+}
 
-	if err := b.initializeMapErrTelemetryMap(m.Maps); err != nil {
-		return err
-	}
-	if err := b.initializeHelperErrTelemetryMap(); err != nil {
-		return err
+func patchConstant(ins asm.Instructions, symbol string, constant int64) error {
+	indices := ins.ReferenceOffsets()[symbol]
+
+	ldDWImm := asm.LoadImmOp(asm.DWord)
+	for _, index := range indices {
+		load := &ins[index]
+		if load.OpCode != ldDWImm {
+			return fmt.Errorf("symbol %v: load: found %v instead of %v", symbol, load.OpCode, ldDWImm)
+		}
+		load.Constant = constant
 	}
+
 	return nil
 }
 
-func (b *EBPFTelemetry) initializeMapErrTelemetryMap(maps []*manager.Map) error {
-	if b.mapErrMap == nil {
-		return nil
+func patchEBPFInstrumentation(m *manager.Manager, bpfTelemetry *EBPFTelemetry, bytecode io.ReaderAt, shouldSkip func(string) bool) error {
+	programs, err := m.GetProgramSpecs()
+	if err != nil {
+		return fmt.Errorf("unable to get program specs: %w", err)
 	}
 
-	z := new(MapErrTelemetry)
-	h := keyHash()
-	for _, m := range maps {
-		// Some maps, such as the telemetry maps, are
-		// redefined in multiple programs.
-		if _, ok := b.mapKeys[m.Name]; ok {
+	return PatchEBPFInstrumentation(programs, bpfTelemetry, bytecode, shouldSkip)
+}
+
+// PatchEBPFInstrumentation accepts eBPF bytecode and patches in the eBPF instrumentation
+func PatchEBPFInstrumentation(programs map[string]*ebpf.ProgramSpec, bpfTelemetry *EBPFTelemetry, bytecode io.ReaderAt, shouldSkip func(string) bool) error {
```

Comment:
Does it make sense to do all programs at once, or would it possible simplify things to operate on a single function at once? That way the skipping could happen outside, by *not* calling this function?

Reply:
That would introduce two problems:
1. When loading programs without the ebpf-manager we still want to perform the patching, for example in the verifier statistics calculator. It would introduce more logic in those code paths.
2. When skipped we still patch the `call -1` with a NOP. We would still need a codepath outside this function for doing that.

IMO it would not result in much simplification because of these reasons.